plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java

```diff
@@ -1237,24 +1332,56 @@ public Optional<ConnectorOutputMetadata> finishCreateTable(
+            }
+            else {
+                TrinoFileSystem fileSystem = fileSystemFactory.create(session);
+                commitVersion = getMandatoryCurrentVersion(fileSystem, handle.getLocation(), handle.getReadVersion().getAsLong()) + 1;
+                if (commitVersion != handle.getReadVersion().getAsLong() + 1) {
+                    throw new TransactionConflictException(format("Conflicting concurrent writes found. Expected transaction log version: %s, actual version: %s",
+                            handle.getReadVersion().getAsLong(),
+                            commitVersion - 1));
+                }
+                transactionLogWriter = transactionLogWriterFactory.newWriter(session, handle.getLocation());
+            }
             appendTableEntries(
-                    0,
+                    commitVersion,
                     transactionLogWriter,
                     randomUUID().toString(),
                     schemaString,
                     handle.getPartitionedBy(),
                     configurationForNewTable(handle.getCheckpointInterval(), handle.getChangeDataFeedEnabled(), columnMappingMode, handle.getMaxColumnId()),
-                    CREATE_TABLE_AS_OPERATION,
+                    handle.isReplace() ? CREATE_OR_REPLACE_TABLE_AS_OPERATION : CREATE_TABLE_AS_OPERATION,
                     session,
                     handle.getComment(),
                     handle.getProtocolEntry());
             appendAddFileEntries(transactionLogWriter, dataFileInfos, physicalPartitionNames, columnNames, true);
+            if (handle.getReadVersion().isPresent()) {
+                long writeTimestamp = Instant.now().toEpochMilli();
+                DeltaLakeTableHandle deltaLakeTableHandle = (DeltaLakeTableHandle) getTableHandle(session, schemaTableName);
+                try (Stream<AddFileEntry> activeFiles = transactionLogAccess.getActiveFiles(
+                        session,
+                        getSnapshot(session, deltaLakeTableHandle),
+                        deltaLakeTableHandle.getMetadataEntry(),
+                        deltaLakeTableHandle.getProtocolEntry(),
+                        deltaLakeTableHandle.getEnforcedPartitionConstraint(),
+                        deltaLakeTableHandle.getProjectedColumns().orElse(ImmutableSet.of()))) {
+                    Iterator<AddFileEntry> addFileEntryIterator = activeFiles.iterator();
+                    while (addFileEntryIterator.hasNext()) {
+                        AddFileEntry addFileEntry = addFileEntryIterator.next();
+                        transactionLogWriter.appendRemoveFileEntry(new RemoveFileEntry(addFileEntry.getPath(), addFileEntry.getPartitionValues(), writeTimestamp, true));
+                    }
+                }
+            }
             transactionLogWriter.flush();
+            writeCommitted = true;
```

Comment:
We need to write a checkpoint here if needed, as in finishInsert
https://github.com/trinodb/trino/blob/d8b57c1f485d015a131e7d9a6b2b6099139b212b/plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java#L1801
If not, running CREATE OR REPLACE repeatedly will eventually become uncomfortable.

Reply:
That's our plan to fix them as a follow-up to this ticket.